### PR TITLE
[FLINK-25994] Implement FileStoreExpire

### PR DESCRIPTION
**Pull request by tsreaper**
Currently FileStoreExpire does not have an implementation. We need an implementation to clean up old snapshots and related files.

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+ * snapshots.
+ */
+public class FileStoreExpireImpl implements FileStoreExpire {
+
+    private final int numRetained;
+    private final FileStorePathFactory pathFactory;
+    private final ManifestList manifestList;
+    private final FileStoreScan scan;
+
+    public FileStoreExpireImpl(
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java:48]: I am thinking that we can provide three options:
- snapshot-expire.retention: retention time for snapshot changes
- snapshot-expire.num-retained
- snapshot-expire.check-interval

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+        for (long id = firstSnapshotIdToExpire; id <= lastSnapshotIdToExpire; id++) {
+            expire(id);
+        }
+    }
+
+    private void expire(long snapshotId) {
+        // delete all sst files removed in the snapshot to expire
+        Map<BinaryRowData, Map<Integer, SstPathFactory>> sstPathFactories = new HashMap<>();
+        for (ManifestEntry entry : scan.withSnapshot(snapshotId).plan().deletedFiles()) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java:96]: Consider upgrade files?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreScan.java [hunk 3]
```diff
@@ -56,5 +56,11 @@
 
         /** Result {@link ManifestEntry} files. */
         List<ManifestEntry> files();
+
+        /**
+         * Deleted sst files in current snapshot. They can be safely deleted from file system if
+         * this snapshot expires.
+         */
+        List<ManifestEntry> deletedFiles();
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreScan.java:64]: This `deletedFiles` is misunderstanding.
I prefer to add a new method `planChangeFiles()`, it returns all files. The caller can compute physical deleted files.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java (hunk 1 above)
Comment (tsreaper) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java:48]: `Expire` is an operation called by users (for example by connector sinks). `snapshot-expire.check-interval` should be checked by users, not by the operation itself.

As for retention time I'll implement this.

Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java:48]: Can you modify the caller too? `StoreGlobalCommitter`?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    @Override
+    public void expire() {
+        Long latestSnapshotId = pathFactory.latestSnapshotId();
+        if (latestSnapshotId == null) {
+            // no snapshot, nothing to expire
+            return;
+        }
+
+        // binary search for the last snapshot to expire
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java:72]: It is not cost effective to use dichotomous lookups (assuming there are enough surviving snapshots).
Just start traversing from the beginning.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    private void expire(long snapshotId) {
+        Snapshot toExpire = Snapshot.fromPath(pathFactory.toSnapshotPath(snapshotId));
+        Snapshot nextSnapshot = Snapshot.fromPath(pathFactory.toSnapshotPath(snapshotId + 1));
+
+        // if sst file in only used in snapshot to expire but not in next snapshot we can delete it
+        // because each sst file will only be added and deleted once
+        Set<Path> sstInUse = new HashSet<>();
+        FileStorePathFactory.SstPathFactoryCache sstPathFactoryCache =
+                new FileStorePathFactory.SstPathFactoryCache(pathFactory);
+        for (ManifestEntry entry : scan.withSnapshot(nextSnapshot.id()).plan().files()) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java:126]: I can see large number of repeat reads. We can add a TODO to add manifest cache in scan.

Comment (tsreaper) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreExpireImpl.java:126]: We can optimize this by expiring all snapshots at once so no repeated reads.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**